climada/engine/impact.py

```diff
@@ -940,6 +946,172 @@ def _cen_return_imp(imp, freq, imp_th, return_periods):
+
+        """
+
+        per_event_attrs = []
+        nb_events = len(event_obj.event_id)
+        # select the attributes from hazard object with a length or a
+        # row-dimension equal to the number of event_ids
+        for attr, value in event_obj.__dict__.items():
+            try:
+                if len(value) == nb_events:
+                    per_event_attrs.append(attr)
+            except TypeError:
+                try:
+                    if nb_events in value.shape:
+                        per_event_attrs.append(attr)
+                except AttributeError:
+                    pass
+        return per_event_attrs
+
+
+    def select(self,
+               event_ids=None, event_names=None, dates=None,
+               coord_exp=None):
+        """
+        Select a subset of events and/or exposure points from the impact.
+        If multiple variables are set, it returns all the impacts matching
+        at least one of the conditions.
+        
+        Note: the frequencies are NOT adjusted. Method to adjust frequencies
+        and obtain correct eai_exp:
+            1- Select subset of impact according to your choice
+            imp = impact.select(...)
+            2- Adjust manually the frequency of the subset of impact
+            imp.frequency = [...]
+            3- Use select without arguments to select all events and recompute
+            the eai_exp with the updated frequencies.
+            imp = imp.select()
+
+        Parameters
+        ----------
+        event_ids : list[int], optional
+            Selection of events by their id. The default is None.
+        event_names : list[str], optional
+            Selection of events by their name. The default is None.
+        dates : tuple(), optional
+            Tuple of dates to select the events in between. The default is None.
+        coord_exp : (np.ndarray), optional]
+            Selection of exposures coordinates [lat, lon] (in degrees)
+            The default is None.
+
+        Raises
+        ------
+        ValueError
+            If the impact matrix is missing, the eai_exp and aai_agg cannot
+            be updated for a selection of events and/or exposures.
+
+        Returns
+        -------
+        imp : Impact()
+            A new impact object with a selection of events and/or exposures
+
+        """
+
+        if self.imp_mat.shape == (0, 0):
+            raise ValueError("The impact matrix is missing. eai_exp and\
+aai_agg cannot be computed. Recomputed impact.calc with save_mat=True")
+
+        imp = copy.deepcopy(self)
+        nb_events = imp.event_id.size
+        nb_exp = len(imp.coord_exp)
+        
+        sel_exp = np.arange(nb_exp)
+        # filter exposures
+        if isinstance(coord_exp, np.ndarray):
+            sel_exp = [j
+                       for j, coord in enumerate(imp.coord_exp)
+                       if coord in coord_exp
+                      ]
```

Comment:
But this is a slightly different case right? Now it only checks whether it is None, but does not make sure it is an array. Is this better?

Reply:
Yes, this is different. But then it must not be an ndarray. The only occurrence of `coord_exp` is here and in context of the `in` operator.\
Hence it can be any kind of sequence: a string, a list, an array, a set, a dict, a generator, an iterator etc.\
If it's not such a thing the in operator will simply throw an exception.\
But unlike in the original version, it will not be ignored! Well, yes, I think that's preferable.

Reply:
Ok, good point. It should indeed not be ignored if it is not an iterable.